gamemodes/terrortown/entities/entities/ttt_smoke.lua

```diff
@@ -0,0 +1,428 @@
+	self:NetworkVar("Float", 3, "DeactivateTime")
+end
+
+---
+-- @realm shared
+function ENT:Initialize()
+	if self.DebugViz then
+		self:SetMaterial("models/shiny.vtf")
+		self:SetColor( Color(255, 200, 0, 128) )
+		self:SetRenderMode( RENDERMODE_TRANSCOLOR )
+		-- self:SetNoDraw(false)
+		-- self:DrawShadow(true)
+	else
+		self.RenderGroup = RENDERGROUP_TRANSLUCENT
+		self:DrawShadow(false)
+		self:SetNoDraw(false)
+		self:SetRenderMode( RENDERMODE_TRANSCOLOR )
+	end
+
+	--[[ 4: continue as normal ]]--
+	if SERVER then
+		self.LoopSound = CreateSound(self, self.SmokeEmissionSound) --[[@as CSoundPatch]]
+		self.LoopSound:PlayEx(0.125, 32)
+		self:SetCreateTime(CurTime())
+		self:SetAlive(true)
+	end
+	if CLIENT then
+		self:Setup()
+		if self.DebugViz then return end
+		self:ClientEffect()
+	end
+end
+
+---
+-- We use this as a way to setup the RandomWrap function inside of a predicted hook. (Initialize)
+-- This ensures that CurTime() is the same for server/client, and therefore util.SharedRandom will be identical.
+-- @realm client
+function ENT:Setup()
+	self.seed = tostring(CurTime())
+	local call = 0
+	self.RandomWrap = function(min_value, max_value)
+		local value = util.SharedRandom(self.seed, min_value, max_value, call)
+		call = call + 1
+		return value
+	end
+end
+
+---
+-- @realm shared
+function ENT:Think()
+	-- TODO: figure out if this is relevant
+	if ( CLIENT ) then
+		local physobj = self:GetPhysicsObject()
+
+		if ( IsValid( physobj ) ) then
+			physobj:SetPos( self:GetPos() )
+			physobj:SetAngles( self:GetAngles() )
+		end
+	end
+
+	if self:GetExtinguishFires() then
+		game_effects.ExtinguishInRadius(self:GetPos(), self:GetRadius())
+	end
+
+	if self:GetAlive() and CurTime() > (self:GetCreateTime() + self:GetLifeTime()) then
+		self:Disipate()
+	end
+	if SERVER and not self:GetAlive() and CurTime() > (self:GetDeactivateTime() + self:GetFadeTime()) then
+		self:Remove()
+	end
+end
+
+---
+-- Create a color from the smoke grenade.
+-- @todo this is called inside game_effects.SpawnSmoke which is SERVER only, this does not seem correct.
+-- @realm client
+function ENT:GenerateSmokeColor()
+	local RandomWrap = self.RandomWrap
+	local color_override = nil
+
+	if self:GetRandomColor() then
+		local colors_available = table.GetKeys(self.SmokeColors)
+		local color_dex = math.max(1, math.ceil(RandomWrap(0, #colors_available)))
+		local color_name = colors_available[color_dex]
+		print("picked color", color_name, "or", color_dex, "out of", #colors_available)
+		color_override = self.SmokeColors[ color_name ]
+	end
+
+	local mix = math.Clamp(RandomWrap(
+		(self:GetColorBase() - self:GetColorVariance()) / 255,
+		(self:GetColorBase() + self:GetColorVariance()) / 255
+	), 0, 1)
+
+	if color_override ~= nil then
+		local h, s = ColorToHSV( color_override )
+		color_override = HSVToColor(h, s, mix)
+	else
+		color_override = HSVToColor(0, 0, mix)
+		-- legacy:
+		-- local unmix = math.Round(mix*255)
+		-- Color(unmix, unmix, unmix)
```

Comment:
is this comment just for you or does it have any importance?

Reply:
i converted the old method of color choice as a backup before i was confident that the new method would be good; was considering adding a switch for it. as it stands i think the new method is good, so i'm gonna cut these out